nilearn/decomposition/dict_learning.py

```diff
@@ -0,0 +1,275 @@
+                            n_jobs=self.n_jobs,
+                            verbose=self.verbose
+                            )
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", UserWarning)
+                # We use protected function _raw_fit as data
+                # has already been unmasked
+                canica._raw_fit(data)
+            components = canica.components_
+        S = (components ** 2).sum(axis=1)
+        S[S == 0] = 1
+        components /= S[:, np.newaxis]
+        self.components_init_ = components
+
+    def _init_loadings(self, data):
+        self._loadings_init = self._cache(_compute_loadings,
+                                          func_memory_level=2)(
+            self.components_init_,
+            data)
+
+    def fit(self, imgs, y=None, confounds=None):
+        """Compute the mask and the ICA maps across subjects
+
+        Parameters
+        ----------
+        imgs: list of Niimg-like objects
+            See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+            Data on which PCA must be calculated. If this is a list,
+            the affine is considered the same for all.
+
+        confounds: CSV file path or 2D matrixf
+            This parameter is passed to nilearn.signal.clean. Please see the
+            related documentation for details
+        """
+        # Base logic for decomposition estimators
+        BaseDecomposition.fit(self, imgs)
+
+        if self.verbose:
+            print('[DictLearning] Loading data')
+        data = mask_and_reduce(self.masker_, imgs, confounds,
+                               reduction_ratio=self.reduction_ratio,
+                               n_components=self.n_components,
+                               random_state=self.random_state,
+                               memory_level=max(0, self.memory_level - 1),
+                               n_jobs=self.n_jobs,
+                               memory=self.memory)
+        if self.verbose:
+            print('[DictLearning] Learning initial components')
+        self._init_dict(data)
+
+        self._raw_fit(data)
+
+    def _raw_fit(self, data):
+        """Compute the mask and the maps across subjects, using raw_data. Can
+        only be called directly is dict_init and mask_img, or
+        components_init_ is provided
+
+        Parameters
+        ----------
+        data: ndarray,
+            Shape (n_samples, n_features)
+        """
+        n_samples, n_features = data.shape
+
+        if self.verbose:
+            print('[DictLearning] Computing initial loadings')
+        self._init_loadings(data)
+
+        dict_init = self._loadings_init
+
+        n_iter = ((n_features - 1) // 20 + 1) * self.n_epochs
+
+        if self.verbose:
+            print('[DictLearning] Learning dictionary')
+        self.components_, dictionary = self._cache(dict_learning_online,
```

Comment:
Feature request: add an option for fixed dictionaries, that is, only learn the sparse code.

This appears to be more work, since the whole class hinges on dict_learning_online from sklearn.